### PR TITLE
feat(0020): add node rebalance EIP

### PR DESCRIPTION
**Pull request by savonarola**

### CODE REVIEW

File: active/0020-node-evacuation.md [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+```
+
+Request:
+```
+GET /api/v4/node_eviction/heath_check
+```
+
+Response:
+```
+HTTP/1.1 503 OK
```
Comment (thalesmg) [inline active/0020-node-evacuation.md:207]:
```suggestion
HTTP/1.1 503 Service Unavailable
```
?

File: active/0020-node-evacuation.md [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+The internal API behavior is the following:
+
+```erlang
+disabled = emqx_eviction_agent:status().
+```
+
+##### Enable
+
+Enabling Eviction Agent requires two arguments:
+* `kind` label.
```
Comment (thalesmg) [inline active/0020-node-evacuation.md:161]: Q: can it be anything, or is it an enumeration of choices?

File: active/0020-node-evacuation.md [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+
+![Evacuation Coordinator Statuses](0020-assets/evacuation-coordinator-statuses-enforce.png)
+
+#### Algorithm
+
+##### Starting
+
+At the start, the Evacuation Coordinator reads the settings from the local file.
+* If evacuation is enabled Evacuation Coordinator tries to pass to the `evicting` state. It enables the local Eviction Agent.
+    * If the local Eviction Agent is successfully enabled, then the Evacuation Coordinator passed to the `evicting` state.
```
Comment (thalesmg) [inline active/0020-node-evacuation.md:291]:
```suggestion
    * If the local Eviction Agent is successfully enabled, then the Evacuation Coordinator passes to the `evicting` state.
```

File: active/0020-node-evacuation.md [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+* preserves settings in the local file;
+* evacuation Coordinator enables local Eviction Agent;
+* passes to the evicting state.
+
+The internal API is:
+```erlang
+ok = emqx_node_rebalance_evacuation:start(
+    #{
+        conn_evict_rate => 10,
+        server_reference => <<"srv:123">>
```
Comment (thalesmg) [inline active/0020-node-evacuation.md:259]: Q: since the server reference can be a list of servers, maybe this could be a list as well?

**Review by thalesmg — DISMISSED** [review]

File: active/0020-node-evacuation.md (hunk 2 above)
Comment (savonarola) [reply active/0020-node-evacuation.md:161]: Any atom that makes sence only for the client of the API.

File: active/0020-node-evacuation.md (hunk 4 above)
Comment (savonarola) [reply active/0020-node-evacuation.md:259]: MQTT doesn't specify exact Server-Reference format, so finally, I decided not to complexify things and treat it as an opaque value.

File: active/0020-node-evacuation.md [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+the session is kept in the memory of the last node to which the client was connected.
+
+So if a node shuts down, sessions of the clients currently or recently connected to it are lost.
+
+Node evacuation should transfer the maximum possible amount of connections and sessions to the other nodes.
+
+Node rebalancing should transfer a calculated (balanced among the nodes) number of connections and sessions from higher loaded nodes to other nodes.
+
+For evacuation, after it's completed in one node, the node can be shut down by sysadmin (but not automatically).
+When the cluster reaches balance, all nodes should exit the connection prohibiting state for rebalancing.
```
[the remark this answers is not in the document]
Comment (savonarola) [reply active/0020-node-evacuation.md]: Tried to clarify  a bit: https://github.com/emqx/eip/pull/64/commits/237a99857ea836e9c2fc34ef172e125d8b1e528f
Is the description clearer now? 😅

File: active/0020-node-evacuation.md [hunk 6]
```diff
@@ -0,0 +1,401 @@
+# Node evacuation/rebalancing in EMQX 4.x
```
Comment (zmstone) [inline active/0020-node-evacuation.md:1]:
```suggestion
# Load rebalancing in EMQX 4.x
```
And the file name should be updated too.

File: active/0020-node-evacuation.md [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+
+## Abstract
+
+Node rebalancing is the enforced process of client and session migration from a set of nodes to some other ones.
```
Comment (zmstone) [inline active/0020-node-evacuation.md:5]:
```suggestion
Cluster load rebalancing is the enforced process of client and session migration from a set of nodes to some other ones.
```

File: active/0020-node-evacuation.md [hunk 8]
```diff
@@ -0,0 +1,401 @@
...
+Node evacuation is a special case of rebalancing when we want to migrate all connections and sessions from
+a particular node.
+
+## Motivation
+
+A user might want to shut down arbitrary nodes, e.g., to maintain the underlying
+software or hardware. Or move connections from higher-loaded nodes to lower.
+
+We need to minimize data loss when a node is switched off. The aim of node evacuation
```
Comment (zmstone) [inline active/0020-node-evacuation.md:14]:
```suggestion
We need to minimize data loss when a node is switched off (when data persistence to external DB is not in use). The aim of node evacuation
```

Comment (zmstone) [reply active/0020-node-evacuation.md:14]: Nvm, there is this info below

File: active/0020-node-evacuation.md [hunk 9]
```diff
@@ -0,0 +1,401 @@
...
+
+Response:
+```
+HTTP/1.1 200 OK
+```
+
+The internal API behavior is the following:
+
+```erlang
+disabled = emqx_eviction_agent:status().
```
Comment (zmstone) [inline active/0020-node-evacuation.md:157]: should we consider renaming this module to a more generic one?

File: active/0020-node-evacuation.md [hunk 10]
```diff
@@ -0,0 +1,401 @@
...
+Response:
+```
+HTTP/1.1 200 OK
+
+{"status": "disabled"}
+```
+
+Request:
+```
+GET /api/v4/node_eviction/heath_check
```
Comment (zmstone) [inline active/0020-node-evacuation.md:146]: Should we consider a more generic api name?
```suggestion
GET /api/v4/load_rebalance/availability_check
```

File: active/0020-node-evacuation.md [hunk 11]
```diff
@@ -0,0 +1,401 @@
...
+
+##### `disabled` state
+
+The disabled state is the default. In the `disabled` state Eviction Agent does not influence the node anyhow.
+
+HTTP responses are the following:
+
+Request:
+```
+GET /api/v4/node_eviction/status
```
Comment (zmstone) [inline active/0020-node-evacuation.md:134]:
```suggestion
GET /api/v4/load_rebalance/status
```

File: active/0020-node-evacuation.md (hunk 4 above)
Comment (thalesmg) [reply active/0020-node-evacuation.md:261]: From the spec, I think that it should be a space-separated list of servers.  Each server format is unspecified, but space-separated: https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901255

> The Server Reference is a UTF-8 Encoded String. The value of this string is a space separated list of references. The format of references is not specified here.

**Comment by savonarola** [comment]
>The nodes under maintenance should be actually re-enabled explicitly before starting to take in connections again, right?
Yes, this is true for evacuated ones.

File: active/0020-node-evacuation.md (hunk 10 above)
Comment (savonarola) [reply active/0020-node-evacuation.md]: ✅

**Comment by zmstone** [comment]
maybe the title of this PR and the md filenames both to be renamed to rebalance.

**Comment by savonarola** [comment]
>maybe the title of this PR and the md filenames both to be renamed to rebalance.
✅

**Review by zmstone — APPROVED** [review]

**Pull request merged**